src/clients/lan_auto.rb

```diff
@@ -96,37 +92,19 @@ def main
         # in case keep_install_network is set to true (in AY)
         # we'll keep values from installation
         # and merge with XML data (bnc#712864)
-        if Ops.get_boolean(@param, "keep_install_network", false) == true
+        if @param["keep_install_network"]
           # read settings from installation
           Lan.Read(:cache)
           # export settings into AY map
           @from_system = Lan.Export
-          @dns = Ops.get_map(@from_system, "dns", {})
-          @routing = Ops.get_map(@from_system, "routing", {})
+          @dns = @from_system["dns"] || {}
+          @routing = @from_system["routing"] || {}
 
-          # copy the keys/values that are not existing in the XML
+          # copy the keys/values that are not existing in the XMomm
           # so we merge the inst-sys settings with the XML while XML
           # has higher priority
-          #
-          # bnc#796580 The problem with this is that due to compatibility with
-          # older profiles, a missing element may have a different meaning than
-          # "use what the filesystem/kernel currently uses".
-          # In particular, a missing write_hostname [1] means
-          # "use the product default from DVD1/control.xml".
-          # Other elements may have similar problems,
-          # to be fixed post-PTF for maintenance.
-          Ops.set(@param, "dns", {}) if !Builtins.haskey(@param, "dns")
-          Builtins.foreach(@dns) do |key, value|
-            if !Builtins.haskey(Ops.get_map(@param, "dns", {}), key) &&
-                key != "write_hostname"
-              Builtins.y2milestone(
-                "(dns) taking %1 from inst-sys. Value = %2",
-                key,
-                value
-              )
-              Ops.set(@param, ["dns", key], value)
-            end
-          end
+          @param["dns"] = NetworkAutoYast.instance.merge_dns(@param["dns"] || {}, @dns)
+
           Ops.set(@param, "routing", {}) if !Builtins.haskey(@param, "routing")
```

Comment:
If we're touching `merge_dns` then we should touch the next chunk of code which does a very similar thing.